### ISSUE TITLE
Upstreaming

### ISSUE THREAD

**Issue by Ericson2314** [timbertson/opam2nix-packages#17]
Unless I'm missing something, this already seems far superior to everything that exists so far. When would you consider `opam2nix` and `opam2nix-packages` good enough to merge/move to the NixOS org? I hope the answer is soon. My model for the next step how `haskellngPackages` existed in parallel for a while before the old stuff was removed too.

Relatedly, I hope to work on cross compilation for this, now that nixpkgs has fairly good support for it in general. Meatier changes like that might be easier when nixpkgs and ocaml changes can be done in tandem.

**Comment by timbertson** [comment]
I think it may be good enough to upstream, but to be honest I'm not ready to take on the responsibility of maintaining it as the canonical source of ocaml nix packages. A few things to consider:

 - automated updates (I'm currently doing a weekly travis build, an upstreamed solution would probably need to be fancier and do something like whatever the haskell packages are doing)
 - we should talk to the folks who did the haskell stuff, see if there are any pitfalls we should avoid (both in terms of the migration and in terms of how we structure the packages)
 - once we upstream, changing the interface gets harder and requires more effort. I don't know how many people have tested opam2nix, but as you say it seems to work better than the existing solutions, so perhaps this is not worth worrying about right now.

Basically, I'm reluctant because it feels like a lot of work, and I don't have the time to take that on any time soon. I would like to see it done though, so if you have any ideas or desire to look into the above, I'd be happy to discuss it.

Also, now that overlays exist, I wonder if it would be useful to integrate opam2nix as a third-party overlay first, to help get people trying it out and working out kinks in the design? That seems like a good step, although maybe it's pretty trivial and the real difficulty is getting an approach for both layout and automation that the nixpkgs maintainers are happy with.

**Comment by Ericson2314** [comment]
> I think it may be good enough to upstream, but to be honest I'm not ready to take on the responsibility of maintaining it as the canonical source of ocaml nix packages. 

Ah makes perfect sense. As I see it, there's a way to do a provisional upstream so you have *fewer* obligations. Indeed sharing the burden is sort of the point.

> we should talk to the folks who did the Haskell stuff

As a Haskeller (far more than OCamler) who works on that, I'd like to use this a model for how that ought to work in a few instances :).

> Once we upstream, changing the interface gets harder and requires more effort.

 I hope this can *not* be the case while its in a provisional state. I hope to if anything move faster with all the packages in one repo. (opam2nix itself would remain a separate repo like cabal2nix I presume.)

> Also, now that overlays exist, I wonder if it would be useful to integrate opam2nix as a third-party overlay first, to help get people trying it out and working out kinks in the design? That seems like a good step, although maybe it's pretty trivial and the real difficulty is getting an approach for both layout and automation that the nixpkgs maintainers are happy with.

Yeah so at least with the cross stuff I've done so far, overlays would have been a real hamper because many changes are quite cross cutting. I also agree that the layout and automation is the hard part.

I also wonder if, while this is in a provisional state, we can do more import-from-derivation to avoid committing a ton of files to nixpkgs. IFS is generally not allowed in exposed packages (as opposed to convenience functions like `callCabal2nix`), but hopefully it can be OK for provisional things. This would drastically reduce churn while things are still in flux.

CC @shlevy

**Comment by shlevy** [comment]
@Ericson2314 Is there something specific you want to draw my attention to here? I haven't been following this thread or this project, sorry.

**Comment by Ericson2314** [comment]
@timbertson I talked to @shlevy about just about process in general. Maybe overlays is fine. I suppose you could read this ticket for a request how I could be more involved in general :). I don't have too much time but if there's a sufficiently-bounded amount of work before it is truly ready to be upstreamed and not-provisional, I could pitch in thorough that.

**Comment by timbertson** [comment]
One thing I did just remember is that downloads aren't integrity-checked yet (#18), that should be a blocker for upstreaming (and I ought to have done it before now, it just slipped my mind once everything was working). You'd be welcome to implement that opam2nix itself if you like.

This affects IFD, since `opam2nix generate` currently relies on a persistent cache of downloaded sources and therefore can't efficiently be done as a nix derivation. Aside from that I like your idea of not actually checking in the `repo/packages` directory, since it's a whole lot of noise. Note that we _already_ use IFD (we generate a selections.nix for your chosen set of packages then import that), so we're already dependant on IFD.

I don't know how keen you would be to drive it, but another thing to do is to kick off a discussion on nix-devel (or maybe just a nixpkgs issue) to see what blockers there are before nixpkgs maintainers would be happy to merge opam2nix alongside the existing ocamlPackages. It may be that everyone's fine with giving it a go and there isn't much to discuss, but that seems optimistic ;)
